args/audio/systems/audiosystem.inl

```diff
@@ -0,0 +1,279 @@
+
+		alDopplerFactor(1.0f);
+		alDistanceModel(AL_EXPONENT_DISTANCE);
+
+		queryInformation();
+
+		//m_sourcePositions = std::unordered_map<ecs::component_handle<audio_source>, position>();
+
+		//ARGS function binding
+
+		sourceQuery = createQuery<audio_source>();
+
+		createProcess<&AudioSystem::update>("Audio");
+		bindToEvent<events::component_creation<audio_source>, &AudioSystem::onAudioSourceComponentCreate>();
+		bindToEvent<events::component_destruction<audio_source>, &AudioSystem::onAudioSourceComponentDestroy>();
+		bindToEvent<events::component_creation<audio_listener>, &AudioSystem::onAudioListenerComponentCreate>();
+		bindToEvent<events::component_destruction<audio_listener>, &AudioSystem::onAudioListenerComponentDestroy>();
+
+		// Release context on this thread
+		alcMakeContextCurrent(nullptr);
+	}
+
+	inline void AudioSystem::queryInformation()
+	{
+		const ALchar* vendor = alGetString(AL_VENDOR);
+		const ALchar* version = alGetString(AL_VERSION);
+		const ALchar* renderer = alGetString(AL_RENDERER);
+		const ALchar* openALExtensions = alGetString(AL_EXTENSIONS);
+		const ALchar* ALCExtensions = alcGetString(data::alDevice, ALC_EXTENSIONS);
+		ALCint srate;
+		alcGetIntegerv(data::alDevice, ALC_FREQUENCY, 1, &srate);
+		log::info("OpenAL info:\n\n\t\t\tOpenAL information\n\tVendor: {}\n\tVersion: {}\n\tRenderer: {}\n\tOpenAl Extensions: {}\n\tALC Extensions: {}\n\tDevice samplerate: {}\n",
+			vendor, version, renderer, openALExtensions, ALCExtensions, srate);
+	}
+
+	inline void AudioSystem::update(time::span deltatime)
+	{
+		alcMakeContextCurrent(data::alContext);
+
+		int i;
+
+		for (auto entity : sourceQuery)
+		{
+			auto sourceHandle = entity.get_component_handle<audio_source>();
+
+			audio_source source = sourceHandle.read();
+			position p = entity.read_component<position>();
+			position& previousP = m_sourcePositions.at(sourceHandle);
+			math::vec3 vel = previousP - p;
+			previousP = p;
+			alSource3f(source.m_sourceId, AL_POSITION, p.x, p.y, p.z);
+			alSource3f(source.m_sourceId, AL_VELOCITY, vel.x, vel.y, vel.z);
+			log::debug(source.m_changes);
+			if (source.m_changes & 1)
+			{
+				// Pitch has changed
+				alSourcef(source.m_sourceId, AL_PITCH, source.getPitch());
+			}
+			if (source.m_changes & 2)
+			{
+				// Gain has changed
+				alSourcef(source.m_sourceId, AL_GAIN, source.getGain());
+			}
+
+			source.clearChanges();
+			sourceHandle.write(source);
+		}
+
+		if (m_listenerEnt)
+		{
+			position p = m_listenerEnt.read_component<position>();
+			rotation r = m_listenerEnt.read_component<rotation>();
+
+			setListener(p, r);
+
+			math::vec3 vel = m_listenerPosition - p;
+			m_listenerPosition = p;
+			alListener3f(AL_VELOCITY, vel.x, vel.y, vel.z);
+		}
+
+		alcMakeContextCurrent(nullptr);
+	}
+
+#pragma region Component creation&destruction
+	inline void AudioSystem::onAudioSourceComponentCreate(events::component_creation<audio_source>* event)
+	{
+		alcMakeContextCurrent(data::alContext);
```

Comment:
what if making the context current fails because another thread is utilizing the context?